lib/active_record_host_pool/connection_adapter_mixin.rb

```diff
@@ -3,13 +3,53 @@
 require "active_record/connection_adapters/mysql2_adapter"
 
 module ActiveRecordHostPool
+  module PerThreadHashAccess
+    # PerThreadHashAccess.module_for(...) returns a module which scopes hash accessor methods
+    # to the given connection_id with thread-local values. It uses `super` as a fallback. This is to
+    # make access to the config object on a ActiveRecord::ConnectionAdapters::Mysql2Adapter
+    # thread-safe.
+    #
+    # Because AHRP shares the connection across members of its connection pool it is possible for
+    # multiple threads be simultaneously trying to talk to the database.
+    def self.module_for(connection_id)
+      Module.new do
+        define_method(:[]=) do |key, value|
+          _ahrp_storage[key] = value
+        end
+
+        define_method(:[]) do |key|
+          if _ahrp_storage.key?(key)
+            _ahrp_storage[key]
+          else
+            _ahrp_storage[key] = super(key)
+          end
+        end
+
+        define_method(:fetch) do |*args, &blk|
+          key, = args
+          _ahrp_storage.fetch(key) do
+            _ahrp_storage[key] = super(*args, &blk)
+          end
+        end
+
+        private
+
+        define_method(:_ahrp_storage) do
+          DatabaseSwitch.thread_local_storage[connection_id]
+        end
+      end
+    end
+  end
+
   module DatabaseSwitch
     def self.included(base)
       base.class_eval do
-        attr_reader(:_host_pool_current_database)
+        def _host_pool_current_database
+          _ahrp_storage[:_host_pool_current_database]
+        end
```

Comment:
Any reason to not move the `_host_pool_current_database` accessor methods out of the class_eval and into the main module body?